### PR TITLE
patch/as tactic

### PR DESCRIPTION
**Pull request by randair**
Patch proof tactic of the form "patch old new as hyp_name", which behaves as the Patch Proof command does but creates a hypothesis instead. Common patch functionality is refactored.

### CODE REVIEW

File: plugin/src/patcher.ml4 [hunk 1]
```diff
@@ -103,42 +107,68 @@ let invert_patch n env patch sigma =
...
+let patch env n try_invert a search sigma define =
   let reduce = try_reduce reduce_remove_identities in
   let sigma, patch_to_red = search env a sigma in
   let sigma, patch = reduce env sigma patch_to_red in
   let prefix = Id.to_string n in
-  ignore (define_term n sigma patch false);
   (if !opt_printpatches then
     print_patch env sigma prefix patch
   else
     Printf.printf "Defined %s\n" prefix);
```
Comment (tlringer) [inline plugin/src/patcher.ml4]: I would move this print statement underneath the place you move define_term

Comment (tlringer) [reply plugin/src/patcher.ml4]: So I guess inside of the function you pass to the higher-order function

File: plugin/src/patcher.ml4 [hunk 2]
```diff
@@ -103,42 +107,68 @@ let invert_patch n env patch sigma =
...
-  if try_invert then
+  (if try_invert then
     try
       let inv_n_string = String.concat "_" [prefix; "inv"] in
       let inv_n = Id.of_string inv_n_string in
       invert_patch inv_n env patch sigma
```
Comment (tlringer) [inline plugin/src/patcher.ml4]: I think you will need to play a similar game here and pass `define` to `invert_patch`, then call it similarly

File: plugin/src/patcher.ml4 [hunk 3]
```diff
@@ -103,42 +107,68 @@ let invert_patch n env patch sigma =
...
     with _ ->
       ()
   else
-    ()
+    ());
+  define sigma patch
```
Comment (tlringer) [inline plugin/src/patcher.ml4]: In which case, the `define` call here should go before the call to `invert_patch`

File: plugin/src/patcher.ml4 [hunk 4]
```diff
@@ -103,42 +107,68 @@ let invert_patch n env patch sigma =
...
 
+(* can't move "define" as the first argument to allow partial 
+   application because it messes with the type of "search" *)
```
Comment (tlringer) [inline plugin/src/patcher.ml4]: I would get rid of this comment, and add a comment here explaining this is the tactic version

File: plugin/src/patcher.ml4 [hunk 5]
```diff
@@ -103,42 +107,68 @@ let invert_patch n env patch sigma =
...
+let patch_tactic env n try_invert a search sigma =
+  patch env n try_invert a search sigma 
+    (fun sigma patch ->
+      letin_pat_tac false None (Names.Name n)
+        (sigma, etrmEConstr.of_constr patch) Locusops.nowhere) 
+  
+let patch_command env n try_invert a search sigma =
```
Comment (tlringer) [inline plugin/src/patcher.ml4]: And I'd add a comment here explaining this is the command version

File: plugin/src/patcher.ml4 (hunk 4 above)
Comment (tlringer) [reply plugin/src/patcher.ml4]: Plus maybe an explanation of what is different from the command version

File: plugin/src/patcher.ml4 [hunk 6]
```diff
@@ -103,42 +107,68 @@ let invert_patch n env patch sigma =
...
-  let sigma, (old_term, new_term) = intern_defs env d_old d_new sigma in
+  let sigma, (old_term, new_term) = intern env d_old d_new sigma in
   let sigma, (d, opts) = configure env old_term new_term cut sigma in
   let change = get_change opts in
   let try_invert = not (is_conclusion change || is_hypothesis change) in
   let search _ _ = search_for_patch old_term opts d in
   patch env n try_invert () search sigma
-
+    
+let patch_proof_tactic n d_old d_new =
```
Comment (tlringer) [inline plugin/src/patcher.ml4:166]: Same here, I'd add comments explaining tactic/command version and what the difference is. They can be brief.

**Review by tlringer — CHANGES_REQUESTED** [review]
Glad this is working :). Just a few comments.

In addition, what happens when you run `./build.sh` followed by `./test.sh`? Can you paste the output into here so I can check it?

I would also like to see a few test cases. Those can build on the existing tests. Happy to help with these next meeting if you'd like. Then we can talk about next steps.

**Comment by randair** [comment]
Oops I left a typo on 137. Fixed and will make the remaining requested changes.

**Comment by randair** [comment]
Attached is the output of ./build.sh followed by ./test.sh.
[output.txt](https://github.com/uwplse/PUMPKIN-PATCH/files/4048888/output.txt)

**Comment by tlringer** [comment]
OK, tests look good. Let's take a shot at the PR feedback when you're in, and then we can move on to producing tactics.

File: plugin/src/patcher.ml4 (hunk 2 above)
Comment (tlringer) [reply plugin/src/patcher.ml4]: Then in the tactic you'll need to define both hypotheses

**Review by tlringer — CHANGES_REQUESTED** [review]
For regression tests, extend `Regress.v` with dummy proofs that call your tactic. 

Then also extend `divide.v` with a copy of the proof anywhere we use the patch currently to use your tactic, and same with `IntegersNew.v`. That should be enough to start.

**Review by tlringer — APPROVED** [review]
Looks good. We can talk tomorrow about what's next.

**Pull request merged**